src/logging.rs

```diff
@@ -0,0 +1,64 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use log::{LevelFilter, Metadata, Record};
+use std::ffi::CString;
+use std::os::raw::c_char;
+use std::os::raw::c_int;
+
+pub type LogCallback = unsafe extern "C" fn(level: c_int, msg: *const c_char);
+
+struct LogProxy {
+    callback: LogCallback,
+}
+
+impl log::Log for LogProxy {
+    fn enabled(&self, _metadata: &Metadata) -> bool {
+        true
+    }
+
+    fn log(&self, record: &Record) {
+        if self.enabled(record.metadata()) {
+            let callback: LogCallback = self.callback;
+            let msg = record.args().to_string();
+            let msg_c = CString::new(msg).unwrap();
+            unsafe {
+                callback(record.level() as c_int, msg_c.as_ptr());
+            }
+            // We do not use std::mem::forget(s), so Rust will reclaim the memory
+            // once msg_c gets cleared. The callback should thus make a copy.
+        }
+    }
+
+    fn flush(&self) {}
+}
+
+// Store the logger instance as a mutable nullable static object
+static mut LOGGER: Option<LogProxy> = None;
+
+// Note: this function may only be called once in the lifetime of a program.
```

Comment:
Instead of this comment here, consider returning an error code when setting the callback, such that the calling code can check whether this succeeded or not.

Reply:
We cannot really return an error in a C function. Or do you mean returning -1 or something?

Reply:
The `log::set_logger` call would panic, I suppose, when called twice, and I imagine that the error messages is quite similar to what we'd do.

Reply:
Ok, maybe not as descriptive: `thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: SetLoggerError(())', src\libcore\result.rs:1165:5`

Reply:
Exactly, returning a value to C might be a good idea, at least something like `WGPU_OK` or `WGPU_ERROR` if this exists already.

Reply:
Mmm, I applied this to `wgpu_set_log_level`, but for `wgpu_set_log_callback` I think I prefer to panic instead. Better fail hard (with a good error message) than fail silently because the caller fails to check the return value.

Reply:
Exactly, so the best way to fail in python is to check the result code, and raise an exception ;)

Reply:
Yeah but there is a risk of forgetting to check it. Also because this pattern is not common in the rest of wgpu. When an application calls this function more than once, it does something really wrong, and it deserves a panic ;)